### PR TITLE
Support for admin tab in React

### PR DESCRIPTION
**Pull request by UncleSamSwiss**
**PR Checklist:**

-   [x] Provide a meaningful description to this PR or mention which issues this fixes.
-   [x] Add tests for your change. This includes negative tests (i.e. inputs that need to fail) as well as baseline tests (i.e. how should the directory structure look like?).
-   [x] Run the test suite with `npm test`
-   [x] If there are baseline changes, review them and make a separate commit for them with the comment "accept baselines" if they are desired changes
-   [x] Ensure the project builds with `npm run build`
-   [x] If you added a required option, also add it to the template creation (`.github/create_templates.ts`)
-   [ ] ~~Add a detailed migration description to `docs/updates` explaining what the user needs to do when manually updating an existing project~~ Not required as this is a new feature
-   [x] Add your changes to `CHANGELOG.md` (referencing the migration description and this PR or the issue you fixed)

**Description:**  
Supports choosing to create the optional tab in admin using React.
A mixture of React and regular HTML is possible (both ways).

### CODE REVIEW

File: .github/create_templates.ts [hunk 1]
```diff
@@ -82,6 +84,7 @@ const templates: Record<string, Answers> = {
 		indentation: "Tab",
 		quotes: "double",
 		adminReact: "yes",
+		tabReact: "yes",
```
Comment (AlCalzone) [inline .github/create_templates.ts:87]: Should we really add a tab to every template by default?
Most adapters don't need or use it.

File: src/lib/createAdapter.ts [hunk 2]
```diff
@@ -42,11 +42,13 @@ export function testCondition(
...
 		} else if ("contains" in cond) {
 			return (
+				answers[cond.name] &&
 				(answers[cond.name] as AnswerValue[]).indexOf(cond.contains) >
-				-1
+					-1
 			);
 		} else if ("doesNotContain" in cond) {
 			return (
+				!answers[cond.name] ||
```
Comment (AlCalzone) [inline src/lib/createAdapter.ts:51]: Why is that necessary? It has been working the entire time without it, even with optional questions. (or did we never create a situation where something depends on an optional question?)

File: templates/admin/src/tab-app.tsx_jsx.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
+import { TemplateFunction } from "../../../src/lib/createAdapter";
```
Comment (AlCalzone) [inline templates/admin/src/tab-app.tsx_jsx.ts:1]: Would it make sense to integrate this into component into the tab.tsx.ts?

File: templates/admin/words.js.ts [hunk 4]
```diff
@@ -4,13 +4,16 @@ import { getTranslatedSettings } from "../../src/lib/translation";
...
+	const useAdminReact = answers.adminReact === "yes";
+	const useTabReact = hasTab && answers.tabReact === "yes";
+	const useReact = useAdminReact || useTabReact;
 
 	// Only do the adapter-specific things when not using React
 	const isAdapter = !useReact && answers.features.indexOf("adapter") > -1;
 	const isWidget = answers.features.indexOf("vis") > -1;
 
-	if (useReact && !isWidget) return;
+	if (useAdminReact && (!hasTab || useTabReact) && !isWidget) return;
```
Comment (AlCalzone) [inline templates/admin/words.js.ts:16]: Not sure about this change. Doesn't line 9 and 10 make it so that thee old condition was okay?

File: templates/gulpfile.js.ts [hunk 5]
```diff
@@ -3,8 +3,8 @@ import { TemplateFunction } from "../src/lib/createAdapter";
 export = (answers => {
 
 	const useTypeScript = answers.language === "TypeScript";
-	const useReact = answers.adminReact === "yes";
-	const skipWordsJs = useReact;
+	const hasTab = answers.adminFeatures && answers.adminFeatures.indexOf("tab") > -1;
+	const skipWordsJs = answers.adminReact === "yes" && (!hasTab || answers.tabReact === "yes");
```
Comment (AlCalzone) [inline templates/gulpfile.js.ts:7]: I'd use the same logic as in words.js here.

File: test/create-adapter.test.ts [hunk 6]
```diff
@@ -186,6 +187,17 @@ describe("adapter creation =>", () => {
...
 
+			it("Adapter, Tab, TypeScript React", async () => {
+				const answers: Answers = {
+					...baseAnswers,
+					adminFeatures: ["tab"],
+					es6class: "yes",
+					adminReact: "yes",
+					tabReact: "yes",
+				};
+				await expectSuccess("adapter_tab_TS_React", answers);
```
Comment (AlCalzone) [inline test/create-adapter.test.ts]: Since this test is aimed at checking the differences `tabReact` introduces, I'd add a file filter here.

**Review by AlCalzone — CHANGES_REQUESTED** [review]
I think we should think really hard if it is worth the additional complexity to choose React for the Tab and config UI separately.

File: .github/create_templates.ts (hunk 1 above)
Comment (UncleSamSwiss) [reply .github/create_templates.ts:87]: Good question, fact is, it is already now enabled in the default config (see line 47). The only thing I did, is make sure if we have React for Admin, we also use React for the Tab.

File: src/lib/createAdapter.ts (hunk 2 above)
Comment (UncleSamSwiss) [reply src/lib/createAdapter.ts:51]: The situation is the following: if you have a widget only, it will cause an exception as the answer for adminFeatures doesn't exist, but it needs to check if it should ask for TabReact. This happened in the unit tests. You can try to revert those lines and run the unit tests, you should get the exception.

File: templates/admin/src/tab-app.tsx_jsx.ts (hunk 3 above)
Comment (UncleSamSwiss) [reply templates/admin/src/tab-app.tsx_jsx.ts:1]: We could do this, I simply kept the same architecture as we have for index.tsx and app.tsx. Which is also the architecture suggested by adapter-react.

Another question is, whether we should switch to functional components. I use them now in ham and it seems easier to write code.

File: templates/admin/words.js.ts (hunk 4 above)
Comment (UncleSamSwiss) [reply templates/admin/words.js.ts:16]: The problem is: useTabReact wil be false even if tab was not selected. 
We need to support the following cases:
- only Admin; use React for Admin (--> useAdminReact=true, hasTab=false, useTabReact=false) leads to no words.js
- Admin React and Tab React (--> useAdminReact=true, hasTab=true, useTabReact=true) leads to no words.js
- Admin React and Tab HTML (--> useAdminReact=true, hasTab=true, useTabReact=false) leads to a words.js
- Admin HTML and Tab React (--> useAdminReact=false, hasTab=true, useTabReact=true) leasds to a words.js
- only Admin; use HTML for Admin (--> useAdminReact=false, hasTab=false, useTabReact=false) leads to a words.js

File: templates/gulpfile.js.ts (hunk 5 above)
Comment (UncleSamSwiss) [reply templates/gulpfile.js.ts:7]: It is, except for the missing isWidget test which wasn't there before. Not sure if we should add this.

File: test/create-adapter.test.ts (hunk 6 above)
Comment (UncleSamSwiss) [reply test/create-adapter.test.ts]: Not sure what you mean by this. How do I add a file filter?

File: .github/create_templates.ts (hunk 1 above)
Comment (AlCalzone) [reply .github/create_templates.ts:87]: You're right, I forgot about that.

File: src/lib/createAdapter.ts (hunk 2 above)
Comment (AlCalzone) [reply src/lib/createAdapter.ts:51]: I see - I probably never ran into this situation without also having a condition for "is adapter".

File: templates/gulpfile.js.ts (hunk 5 above)
Comment (AlCalzone) [reply templates/gulpfile.js.ts:7]: On second look I'm okay with this.

File: test/create-adapter.test.ts (hunk 6 above)
Comment (AlCalzone) [reply test/create-adapter.test.ts]: Here's an example:
https://github.com/ioBroker/create-adapter/blob/c3441f28de83f8c9ca15b53497dfdc771db3777e/test/create-adapter.test.ts#L373-L386
This only outputs the files and directories that are interesting for this test. Usually makes sense for smaller features.

Comment (UncleSamSwiss) [reply test/create-adapter.test.ts]: Updated in 1125166, and added one more test for mixed Admin in HTML and Tab in React, which seems to be a common use case.
